### PR TITLE
Resource update mechanism based on Discover flow and UnifiedResources UI

### PR DESCRIPTION
**Pull request by flyinghermit**
- The action button for SAML apps is updated with `ButtonWithMenu` button. The menu is configured with edit menu button, which opens update dialogue, passing identifier of SAMl application which is to be updated. 

- Discover flow is updated to accommodate update flow.
A new props `updateFlow` have been added of type `DiscoverUpdateProps`:

```
export type DiscoverUpdateProps = {
  // agentMeta specifies ResourceSpec which should be used to
  // start a Discover flow.
  resourceSpec: ResourceSpec;
  // agentMeta includes data that will be used to prepopulate input fields
  // in the respective Discover compnents.
  agentMeta: AgentMeta;
};
```

Implemented to support SAML application update flow - https://github.com/gravitational/teleport.e/issues/4458

### CODE REVIEW

**Review by avatus — APPROVED** [review]

File: web/packages/teleport/src/Discover/useDiscover.tsx [hunk 1]
```diff
@@ -90,11 +94,22 @@ type CustomEventInput = {
...
 };
 
+export type DiscoverUpdateProps = {
+  // resourceSpecForUpdate specifies ResourceSpec which should be used to
+  // start a Discover flow.
+  resourceSpec: ResourceSpec;
+  // agentMetaForUpdate includes data that will be used to prepopulate input fields
+  // in the respective Discover compnents.
+  agentMeta: AgentMeta;
+};
```
Comment (gzdunek) [inline web/packages/teleport/src/Discover/useDiscover.tsx:104]: The field names in comments don't match the actual field names.

File: web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx [hunk 2]
```diff
@@ -177,19 +206,32 @@ const AppLaunch = ({ app }: { app: App }) => {
...
+  function handleSamlAppEditButtonClick() {
+    setUpdateDialogOpen(true);
+    setResourceSpec({
+      name: name,
+      event: DiscoverEventResource.SamlApplication,
+      kind: ResourceKind.SamlApplication,
+      samlMeta: { preset: samlAppPreset },
+      icon: 'Application',
+      keywords: 'saml',
+    });
```
Comment (gzdunek) [inline web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx:218]: Hmm, why do we need two callbacks for that? Doesn't setting resource spec mean that the dialog should be open?
I'd call it `onUpdateResource: (resourceSpec: ResourceSpec) => void`.

File: web/packages/teleport/src/Discover/useDiscover.tsx [hunk 3]
```diff
@@ -455,6 +479,7 @@ export function DiscoverProvider({
     emitErrorEvent,
     emitEvent,
     eventState,
+    isUpdateFlow: Boolean(updateFlow),
```
Comment (gzdunek) [inline web/packages/teleport/src/Discover/useDiscover.tsx:482]:
```suggestion
    isUpdateFlow: !!updateFlow,
```

File: web/packages/teleport/src/Discover/Discover.test.tsx [hunk 4]
```diff
@@ -204,3 +208,91 @@ describe('location state', () => {
...
+    kind: ResourceKind.ConnectMyComputer,
+    event: null,
+    icon: 'Laptop',
+    keywords: '',
+    hasAccess: true,
+  };
+
+  renderUpdate({ resourceSpec: resourceSpec, agentMeta: { resourceName: '' } });
+
+  expect(screen.queryAllByTestId(ResourceKind.Server).length).toBeFalsy();
```
Comment (gzdunek) [inline web/packages/teleport/src/Discover/Discover.test.tsx:271]: This seems to be easier to understand:
```suggestion
  expect(screen.queryByTestId(ResourceKind.Server)).not.toBeInTheDocument();
```
(same for other assertions)

File: web/packages/teleport/src/Discover/useDiscover.tsx [hunk 5]
```diff
@@ -231,6 +247,14 @@ export function DiscoverProvider({
     });
   }
 
+  // trigger update Discover flow.
+  useEffect(() => {
+    if (updateFlow && updateFlow.agentMeta) {
+      updateAgentMeta(updateFlow.agentMeta);
+      onSelectResource(updateFlow.resourceSpec);
+    }
+  }, [updateFlow]);
```
Comment (gzdunek) [inline web/packages/teleport/src/Discover/useDiscover.tsx:256]: This is a react anti pattern https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes.
Maybe we could replace it with setting initial value in `setState`s?

File: web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx [hunk 6]
```diff
@@ -177,19 +206,32 @@ const AppLaunch = ({ app }: { app: App }) => {
...
+      <ButtonWithMenu
+        text="Log In"
         width="123px"
         size="small"
         target="_blank"
         href={samlAppSsoUrl}
         rel="noreferrer"
         textTransform="none"
+        forwardedAs="a"
+        title="Log in to this SAML application"
```
Comment (gzdunek) [inline web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx:231]: I'm not sure if `this` is needed.
```suggestion
        title="Log in to SAML application"
```

File: web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx (hunk 2 above)
Comment (flyinghermit) [reply web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx]: https://github.com/gravitational/teleport/pull/43871/commits/d7c6d0d59d77102a981c11e625dec7ac6aa0d0c5#diff-ffc6bf1267af2b53c1dad2ff62eab6a4b6ae53e7b80ce49e33b62fc5b1808e5e

File: web/packages/teleport/src/Discover/useDiscover.tsx (hunk 5 above)
Comment (flyinghermit) [reply web/packages/teleport/src/Discover/useDiscover.tsx:256]: I am not sure that suggestion fits the bill here.
`onSelectResource` is not a react `SetStateAction` but  a function that starts discover flow. It, among setting up resource view, also dispatches Event, which is an outgoing network request.  Only `updateAgentMeta` directly wraps a SetStateAction.

File: web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx [hunk 7]
```diff
@@ -177,19 +190,31 @@ const AppLaunch = ({ app }: { app: App }) => {
...
+      samlMeta: { preset: samlAppPreset },
+      icon: 'Application',
+      keywords: 'saml',
+    });
+  }
   if (samlApp) {
     return (
-      <ButtonBorder
-        as="a"
+      <ButtonWithMenu
```
Comment (gzdunek) [inline web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx:205]: Shouldn't we show `ButtonBorder` when `setResourceSpec` is empty?
Now when I run OSS, I see an edit button on a SAML app that does nothing.

**Review by gzdunek — APPROVED** [review]

File: web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx (hunk 7 above)
Comment (flyinghermit) [reply web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx]: Now returns `ButtonBorder` if setResourceSpec is empty https://github.com/gravitational/teleport/pull/43871/commits/18a5608fafa9e8423574eaf1537ad39905a9f6fc

**Comment by flyinghermit** [comment]
I slipped https://github.com/gravitational/teleport/pull/43871/commits/0dc5cfbbf3ecf6f12e3b6d4aa3d565a5ad2e7465 after approval, which is needed for the `e` PR https://github.com/gravitational/teleport.e/pull/4584.

**Pull request merged**